### ISSUE TITLE
How to save the embedding matrix on CPU instead of GPU?

### ISSUE THREAD

**Issue by LiuxyEric** [changun/CollMetric#11]
Because the embedding matrix is too large and there are two embedding matrix.
Every time I ran the scripts it return OOM error.
How can I save the embedding matrix on CPU instead of GPU?

**Comment by changun** [comment]
Hi LiuxyEric,

Thank you for submitting the issue. I am wondering which dataset you are
using, because the default dataset should be pretty small and can fit the
most GPU.

You can follow the instructions here to force Tensorflow to run on CPU.
https://github.com/tensorflow/tensorflow/issues/754

**Comment by LiuxyEric** [comment]
@changun Thank u for your advice. I am using my own dataset to train the model. 
It is a very large dataset that contains 10M users.
Any advice for such a large dataset?

**Comment by changun** [comment]
Hi LiuxyEric,

For a large dataset like that (from my experience of dealing with 13M users
dataset, and assuming the # item is much smaller), I found it will be more
efficient to set user vectors as the `average of the item vector` the user
liked, instead of having a randomly initialized vector for each user.

This will require some modification to the CollMetric code, in particular,
you probably will need to create a SparseTensor (BatchSize x # Items) that
contains all the positive items of the users in the user-item pair,
excluding the positive sample item used for gradient descent, and compute
the user vector by `tf.sparse_tensor_dense_matmul(sparse_tensor,
item_embedding)`.

Such an approach will converge much faster, and requires only the GPU
memory that have space to store item embedding and the batch under
computation.

**Comment by LiuxyEric** [comment]
@changun Thank u so much! I'll try to do it.
